content/4_Gold/Tree_Euler.mdx

```diff
@@ -542,6 +563,89 @@ public class Main {
 
 </JavaSection>
 
+<PySection>
+
+```py
+import sys
+
+sys.setrecursionlimit(10**9)
+
+MAX_N = 2 * 10 ** 5
+n, q = map(int, input().split())
+
+v = list(map(int, input().split()))
+
+# BeginCodeSnip{Segment Tree (from PURS module)}
+class SegmentTree:
+	def __init__(self, arr: list[int], n: int):
+		self.arr = arr
+		self.n = n
+		self.tree = [0] * (2 * n)
+		for ii in range(n):
+			self.tree[n + ii] = arr[ii]
+		for ii in reversed(range(n)):
+			self.tree[ii] = self.tree[ii << 1] + self.tree[(ii << 1) | 1]
+
+	def add(self, i: int, v: int):
+		i += self.n
+		self.tree[i] = v
+
+		while i > 1:
+			self.tree[i >> 1] = self.tree[i] + self.tree[i ^ 1]
+			i >>= 1
+
+	def range_sum(self, l: int, r: int) -> int:
+		l += self.n
+		r += self.n
+		sum = 0
+		while l < r:
+			if l & 1:
+				sum += self.tree[l]
+				l += 1
+			if r & 1:
+				r -= 1
+				sum += self.tree[r]
+			l >>= 1
+			r >>= 1
+		return sum
+
+
+# EndCodeSnip
+
+g = [[] for i in range(MAX_N)]
+st = [0 for i in range(MAX_N)]
+en = [0 for i in range(MAX_N)]
+timer = 1
+
+
+def dfs(node: int, parent: int):
+	global timer
+	st[node] = timer
+	timer += 1
+	for son in g[node]:
+		if son != parent:
+			dfs(son, node)
+	en[node] = timer - 1
+
+
+for _ in range(n - 1):
+	x, y = map(int, input().split())
+	x -= 1, y -= 1
```

Comment:
```suggestion
	x -= 1
	y -= 1
```
my guy did you test this

Reply:
Oh, sorry. I thought it works in Python :))